### PR TITLE
[86b1vvvqu] Inventory Management - CRM

### PR DESCRIPTION
**Pull request by wes-sleeman**
# Type

`src`

# Summary

Adds inventory management functions to the CRM library and associated lambdas.

# Changes

* Added product management functions to CRM.
* Added quantity management functions to CRM.
* Added stock & catalogue querying to CRM.
* Added lambda for stock querying and quantity setting.

### CODE REVIEW

**Comment by ZoriTheDesigner** [comment]
Task linked: [CU-86b1vvvqu Add Inventory Management to CRM](https://app.clickup.com/t/86b1vvvqu)

File: src/crm/Crm.fs [hunk 1]
```diff
@@ -110,3 +114,53 @@ let tryFindCustomer (id: CorrelationId) : Customer option =
...
+let catalogue () : Catalogue =
+    genericQuery<Product * Quantity> "Stock" (new QueryFilter()) |> Seq.map fst
+
+let createProduct (product: Product) : Result<Product, string> =
+    let (key, sort) = (product.Name, product.Collection.Name)
+    let filter = QueryFilter("key", QueryOperator.Equal, key)
+    filter.AddCondition("sort", QueryOperator.Equal, sort)
+
+    if genericQuery<Product * Quantity> "Stock" filter |> Seq.isEmpty then
+        genericPut "Stock" { key = key; sort = sort; data = (product, Quantity 0) }
```
Comment (ZoriTheDesigner) [inline src/crm/Crm.fs]: It looks like you are bypassing the validation function `quantity`. Quantities can't be zero, we need to add another case called `OutOfStock` in our domain model.

File: src/crm/Crm.fs [hunk 2]
```diff
@@ -110,3 +114,53 @@ let tryFindCustomer (id: CorrelationId) : Customer option =
...
+let updateProduct (product: Product) : Result<Quantity, string> =
+    let (key, sort) = (product.Name, product.Collection.Name)
+    let filter = QueryFilter("key", QueryOperator.Equal, key)
+    filter.AddCondition("sort", QueryOperator.Equal, sort)
+
+    match genericQuery<Product * Quantity> "Stock" filter |> Seq.tryHead with
+    | Some(_, q) ->
+        genericPut "Stock" { key = key; sort = sort; data = (product, q) }
+        Ok q
+    | None -> Error "Product does not exist."
```
Comment (ZoriTheDesigner) [inline src/crm/Crm.fs]: We need to address the case where there is more than one product with the same name and collection name, it needs to return an error.

File: src/crm/Crm.fs [hunk 3]
```diff
@@ -110,3 +114,53 @@ let tryFindCustomer (id: CorrelationId) : Customer option =
...
+let deleteProduct (product: Product) : Result<Quantity, string> =
+    let (key, sort) = (product.Name, product.Collection.Name)
+    let filter = QueryFilter("key", QueryOperator.Equal, key)
+    filter.AddCondition("sort", QueryOperator.Equal, sort)
+
+    match genericQuery<Product * Quantity> "Stock" filter |> Seq.tryHead with
+    | Some(_, q) ->
+        genericDelete "Stock" { key = key; sort = sort; data = (product, q) }
+        Ok q
+    | None -> Error "Product does not exist."
```
Comment (ZoriTheDesigner) [inline src/crm/Crm.fs]: Same problem as above.

File: src/crm/Crm.fs [hunk 4]
```diff
@@ -110,3 +114,53 @@ let tryFindCustomer (id: CorrelationId) : Customer option =
...
+let updateStockQuantity (product: Product) (quantity: Quantity) : Result<unit, string> =
+    let (key, sort) = (product.Name, product.Collection.Name)
+    let filter = QueryFilter("key", QueryOperator.Equal, key)
+    filter.AddCondition("sort", QueryOperator.Equal, sort)
+
+    match genericQuery<Product * Quantity> "Stock" filter |> Seq.tryHead with
+    | Some(p, _) ->
+        genericPut "Stock" { key = key; sort = sort; data = (p, quantity) }
+        Ok()
+    | None -> Error "Product does not exist."
```
Comment (ZoriTheDesigner) [inline src/crm/Crm.fs]: Same as above.

File: src/lambda/Stock/Function.fs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+[<assembly: LambdaSerializer(typeof<Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer>)>]
+()
+
+type Function() =
+    member __.FunctionHandler (request: APIGatewayHttpApiV2ProxyRequest) (_: ILambdaContext) =
+        try
+            let path = request.RawPath.Trim('/')
+
+            match request.RequestContext.Http.Method.ToUpperInvariant() with
+            | "GET" when path |> String.IsNullOrEmpty ->
```
Comment (ZoriTheDesigner) [inline src/lambda/Stock/Function.fs:23]: Can `path` be null or empty?

File: src/lambda/Stock/Function.fs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                stock ()
+                |> Json.serialize
+                |> fun b -> new APIGatewayHttpApiV2ProxyResponse(StatusCode = int HttpStatusCode.OK, Body = b)
+            | "GET" ->
+                genericQuery<Product * Quantity> "Stock" (new QueryFilter("key", QueryOperator.Equal, path))
```
Comment (ZoriTheDesigner) [inline src/lambda/Stock/Function.fs:28]: `path` is not product name here.

File: src/lambda/Stock/Function.fs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                match request.PathParameters.TryGetValue "q" with
+                | true, v -> Some v
+                | false, _ -> None
+                |> Option.bind (fun s ->
+                    match Int32.TryParse s with
+                    | true, i -> Some i
+                    | false, _ -> None)
+                |> Option.bind quantity
+                |> Option.bind (fun quantity ->
+                    genericQuery<Product * Quantity> "Stock" (new QueryFilter("key", QueryOperator.Equal, path))
```
Comment (ZoriTheDesigner) [inline src/lambda/Stock/Function.fs:42]: Same as above.

**Review by ZoriTheDesigner — CHANGES_REQUESTED** [review]

File: src/lambda/Stock/Function.fs (hunk 5 above)
Comment (wes-sleeman) [reply src/lambda/Stock/Function.fs:23]: Change made. Now it can! :)

**Review by ZoriTheDesigner — COMMENTED** [review]
Looks good.

**Review by ZoriTheDesigner — APPROVED** [review]

**Pull request merged**